text/0000-structural-records.md

```diff
@@ -0,0 +1,1170 @@
+```
+
+We extend the expression grammar with:
+
+```rust
+expr ::= ... | expr_srec ;
+
+expr_srec ::= "{" field_init "," field_inits_tail? "}" ;
+field_inits_tail ::= field_inits ","? | (field_inits ",")? DOTDOT expr ;
+field_inits ::= field_init ("," field_init)* ;
+field_init ::= ident | field ":" expr ;
+```
+
+Note that this grammar permits `{ 0: x, 1: y }`.
+We do so to increase reuse in parsing, to improve consistency,
+and to simplify for macro users.
+This is also applied in the grammars below.
+
+We extend the pattern grammar with:
+
+```rust
+pat ::= ... | pat_srec ;
+
+pat_srec ::= "{" pat_srec_start "}" ;
+pat_srec_start ::= DOTDOT | pat_field "," pat_srec_tail
+pat_srec_tail ::= pat_fields ","? | (pat_fields ",")? DOTDOT? ;
+pat_fields ::= pat_field ("," pat_field)* ;
+pat_field ::= BOX? binding_mode? ident | field ":" pat ;
+```
+
+Finally, we extend the type grammar with:
+
+```rust
+ty ::= ... | ty_srec ;
+
+ty_srec ::= "{" (ty_field ",")+ (ty_field ","?)? "}" ;
+ty_field ::= field ":" ty ;
+```
+
+Post macro expansion, we check that positional and named fields are not mixed.
+We also check that there are no gaps in the positional fields when sorted.
+
+### Static semantics: Type checking and inference
+
+The static semantics of structural records are more or less equivalent to those
+of positional tuples with minor differences, particularly that structural records
+have named fields and tuples don't. Below, we give a partial (because otherwise
+we'd be replicating every aspect of tuples...) overview of the rules.
+
+*Note:* While the syntax in the [grammar] section was the actual concrete
+surface syntax, the syntax in this section is *always* abstract syntax.
+
+Let:
+
++ `n` range over the natural numbers.
++ `i` denote an index over some set or list.
++ `f` range over valid identifiers.
++ `Γ` be a well formed (typing) environment.
++ `σ` range over types.
++ `τ` range over terms.
++ `p` range over patterns.
++ `?x` denote a *fresh* unification variable named `x`.
++ `Γ ⊢ x : σ` denote a judgement that `x` is of type `σ` in the environment `Γ`.
+  You can read this as *"`Γ` entails that `x` is of type `σ`"*.
++ `[x]n` denote `n` amount of `x`s. We write `x_i` for the `i`th `x`.
+
+#### Type identity
+
+Given a structural record type `{ [f_i: τ_i]n }`,
+we consider all permutations of the field-type pairs to be the same type.
+For example, `{ foo: u8, bar: bool }` and `{ bar: bool, foo: u8 }` are
+the same type, but `{ foo: bool, bar: u8 }` is not.
+While this can be seen as a form of subtyping in the type system,
+here we merely see it as part of the concrete syntax which can be done
+away with by sorting on the field names before type checking.
+However, care must be taken to provide error messages that show the types
+as the user wrote them and not as how the type system seems them.
+
+Furthermore, for a structural record to be a well-formed type,
+the component types must be and each field name may only occur once.
```

Comment:
the component types must be what?

Reply:
They must be well-formed. :) I've clarified.